File: src/main/java/qj/codejam/y2008/qualification/a/SavingTheUniverse.java
package qj.codejam.y2008.qualification.a;

import java.util.LinkedList;
import java.util.List;

import qj.codejam.CodeJam;
import qj.util.Cols;

public class SavingTheUniverse {
	public static void main(String[] args) {
//		String resName = "sample.in";
		String resName = "A-large-practice.in";
//		String resName = "A-small-practice.in";
		CodeJam.solver(resName, lineF -> {
			int seCount = Integer.parseInt(lineF.e());
			final LinkedList<String> ses = new LinkedList<String>();
			for (int i = 0; i < seCount; i++) {
				ses.add(lineF.e());
			}
			
			final LinkedList<String> words = new LinkedList<String>();
			int wCount = Integer.parseInt(lineF.e());
			for (int i = 0; i < wCount; i++) {
				words.add(lineF.e());
			}
			
			return () -> "" + howManySwitches(words, ses);
		})
		.solve();
	}


	private static int howFarCanGo(String se, List<String> words) {
		int indexOf = words.indexOf(se);
		if (indexOf==-1) {
			return words.size();
		}
		return indexOf;
	}


	public static int howManySwitches(final List<String> words,
			final List<String> ses) {
		if (words.size() == 0) {
			return 0;
		}
		
		int goSize = Cols.<String>maxI(ses, se -> howFarCanGo(se, words));
		
		if (goSize==words.size()) {
			return 0;
		}
		
		return 1 + howManySwitches(words.subList(goSize, words.size()), ses);
	}
}

File: src/main/scala/qs/codejam/y2008/qualification/a/SavingTheUniverse.scala
package qs.codejam.y2008.qualification.a

import qs.codejam.CodeJam
import qj.util.Cols

/**
 * Created by quan on 1/10/15.
 */
object SavingTheUniverse {

  def main(args: Array[String]) {
  //    val resName: String = "A-large-practice.in"
  //    val resName: String = "A-small-practice.in"
    val resName = "sample.in"


    CodeJam.solver(resName, lineF => {
      val seCount = lineF() toInt
      val ses = (for (i <- 1 to seCount) yield lineF()) toList

      val wCount = lineF() toInt
      val words = (for (i <- 1 to wCount) yield lineF()) toList

      () => {"" + howManySwitches(words, ses)}

    })
    .solve()

  }

  private def howFarCanGo(se: String, words: List[String]): Int = {
    val indexOf: Int = words.indexOf(se)
    if (indexOf == -1) {
      return words.size
    }
    indexOf
  }

  def howManySwitches(words: List[String], ses: List[String]): Int = {
    if (words.size == 0) {
      return 0
    }
    // ((se) => howFarCanGo(se, words))
    val goSize: Int = ses.map(se => howFarCanGo(se, words)).max
    if (goSize == words.size) {
      return 0
    }

    1 + howManySwitches(words.slice(goSize, words.size), ses)
  }

}
